Sources/SwiftDocC/Infrastructure/DocumentationContext.swift

```diff
@@ -2350,6 +2354,100 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
         }
         return automaticallyCuratedSymbols
     }
+
+    /// Handles overloaded symbols by grouping them together into one page.
+    private func handleOverloads(relationships: [SymbolGraph.Relationship]) {
+        guard FeatureFlags.current.isExperimentalOverloadedSymbolPresentationEnabled else {
+            return
+        }
+        
+        var parentToChildren = [SymbolGraph.Symbol.Identifier : [DocumentationNode]]()
+        
+        // Walk the symbol graph relationships and look for parent <-> child links.
+        // Build up the parentToChildren dictionary of all children under each parent.
+        for relationship in relationships {
```

Comment:
It's also worth noting that not all symbol relationships contribute to the link / name hierarchy directly. For example, a "default implementation of" relationship is between the default implementation and the protocol requirement. For example

```swift
protocol SomeProtocol {
  func something() // the "protocol requirement"
}

extension SomeProtocol {
  func something() { // the "default implementation"
    // do something here
  } 
}
```

Adding the default implementation as a child of the protocol requirement would give it a path of "SomeProtocol/something()/something()" but we'd expect both symbols to have a path of "SomeProtocol/something()" which means that the default implementation symbol should be added as a child of the protocol requirement's parent symbol. 

There are a few other relationship kinds that follow this pattern which is all the more reason to avoid building up another symbol hierarchy based on relationships.

Reply:
That's a good point. The link resolver sounds like it would be great to use if it already knows the symbol collisions. Is there a way of using the link resolver to find the symbol collisions before links are actually resolved? I was hoping to recognize overloads and add a new node to the topic graph that represents a group of overloads before links were resolved. That way when writers link to a merged group of overloads (using a link without a hash) the link resolver might be more easily taught to resolve the link because the corresponding node for that link would already exist. However, I'm less familiar with how link resolution works, so I might be making incorrect assumptions about the best way to help resolve overload links without hashes.

Reply:
PS:
I didn't know this 2 years ago but I've l've learnt the hard way by also writing this code and then gradually fixing bugs in it.

Reply:
The link resolver is built up incrementally as more and more content is registered with the documentation context. All the symbol information is added when the symbol graphs are loaded (before the symbols are registered with the context) so if you only need to know about symbols then you can rely on all the symbols and symbol collisions being known.

The core of the link resolution implementation is a tree-like data structure—that's aware of collisions and disambiguation—which is controlled by `PathHierarchy`. Nothing interacts with it directly except for `PathHierarchyBasedLinkResolver` which provides an API to the rest of DocC and translates between `ResolvedTopicReference` values and the path hierarchy's internal identifiers. 

The "link resolver" is mainly responsible for finding what page a given link refers to (PathHierarchy+Find.swift) but since it knowns the most about collisions and disambiguation it also determines the paths of all the symbol pages (PathHierarchy+DisambiguatedPaths.swift).

To separate responsibilities, the "link resolver" (referring to `PathHierarchyBasedLinkResolver` and everything it wraps) doesn't call out to the rest of DocC, for example update the topic graph. Instead it provides API that provide the data that other parts of DocC need to make those modifications. For example: `traverseSymbolAndParentPairs(_:)` is used to do auto-curation.

If you want to perform some action on all the symbols that have symbol kind collisions you could possibly add something like this to `PathHierarchyBasedLinkResolver `:

```swift
func traverseSymbolAndSomething(_ observe: (_ symbol: ResolvedTopicReference, _ something: [ResolvedTopicReference]) -> Void) {
    for (id, node) in pathHierarchy.lookup {
        guard node.symbol != nil, let reference = resolvedReferenceMap[id] else { continue }
        
        // This is peeking into implementation details of how the path hierarchy internally stores disambiguation.
        // Once we know what data and what structure we need we should probably create a PathHierarchy API that hides those implementation details.
        for (_ /*name*/, disambiguation) in node.children {
            for (_ /*kind*/, innerStorage) in disambiguation.storage where innerStorage.count > 1 {
                let childrenWithSameKind = innerStorage.values.compactMap {
                    // Filter out any non-symbols and pages that don't have a known reference
                    $0.symbol == nil ? nil : resolvedReferenceMap[$0.identifier]
                }
                // ???: Does this need the parent reference or just the children? Does it need the name and/or kind?
                observe(reference, childrenWithSameKind)
            }
        }
    }
}
```

That said, at some point the core implementation will need to have some awareness of these combined overload pages so that that can be taken into account when resolving links. 

The code above is probably a decent starting point but I would expect it to change once the core implementation knowns more about combined overload pages.